### ISSUE TITLE
BBC: Blead Breaks Hash::MoreUtils

### ISSUE THREAD

**Issue by cjg-cguevara** [Perl/perl5#20132]
This is a bug report for perl from "Carlos Guevara" <carlos@carlosguevara.com>,
generated with the help of perlbug 1.42 running under perl 5.37.3.

-----------------------------------------------------------------
[Please describe your issue here]

BBC: Blead Breaks Hash::MoreUtils

Please see http://matrix.cpantesters.org/?dist=Hash::MoreUtils

[Please do not change anything below this line]
-----------------------------------------------------------------
---
Flags:
    category=core
    severity=low
---
Site configuration information for perl 5.37.3:

Configured by cpan at Thu Aug 18 11:49:39 EDT 2022.

Summary of my perl5 (revision 5 version 37 subversion 3) configuration:
  Commit id: 52ed379edc11b00708b403ff03aaf5e3a1a2eb81
  Platform:
    osname=linux
    osvers=5.15.60-0-lts
    archname=x86_64-linux
    uname='linux cjg-alpine3 5.15.60-0-lts #1-alpine smp tue, 16 aug 2022 09:10:36 +0000 x86_64 linux '
    config_args='-des -Dprefix=~/bin/perl-blead -Dscriptdir=~/bin/perl-blead/bin -Dusedevel -Duse64bitall'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='cc'
    ccflags ='-D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    optimize='-O2'
    cppflags='-D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong'
    ccversion=''
    gccversion='10.3.1 20211027'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/lib /usr/local/lib /lib
    libs=-lpthread -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -ldl -lm -lcrypt -lutil -lc
    libc=/usr/lib/libc.a
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector-strong'

---
@INC for perl 5.37.3:
    /home/cpan/bin/perl-blead/lib/site_perl/5.37.3/x86_64-linux
    /home/cpan/bin/perl-blead/lib/site_perl/5.37.3
    /home/cpan/bin/perl-blead/lib/5.37.3/x86_64-linux
    /home/cpan/bin/perl-blead/lib/5.37.3

---
Environment for perl 5.37.3:
    HOME=/home/cpan
    LANG=C.UTF-8
    LANGUAGE (unset)
    LC_COLLATE=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/cpan/bin/perl-blead/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

**Comment by jkeenan** [comment]
Bisection points to this commit:
```
9fdd7fc4796d89d16dceea42f2af91e4fde296ed is the first bad commit
commit 9fdd7fc4796d89d16dceea42f2af91e4fde296ed
Author: Richard Leach <richardleach@users.noreply.github.com>
Date:   Fri Jul 8 14:52:19 2022 +0000
```
... which means this has the same breakage point as the modules cited in https://github.com/Perl/perl5/issues/20114.
@richardleach, can you take a look?

Given the amount of downstream breakage this change has generated, should we consider reverting it? (Or is it simply exposing dodgy code downstream?)

**Comment by jkeenan** [comment]
Failures look like this:
```
[perlmonger: Hash-MoreUtils-0.06-0] $ bleadprove -vb t/01-hash.t 
t/01-hash.t .. 
ok 1 - simple slice
ok 2 - slice to nonexistent
...
ok 57 - slice_grep_map on values
ok 58 - slice_grep_map use slice_grep
Can't use an undefined value as a HASH reference at /usr/home/jkeenan/.cpan/build/Hash-MoreUtils-0.06-0/blib/lib/Hash/MoreUtils.pm line 334.
# Tests were run but no plan was declared and done_testing() was not seen.
# Looks like your test exited with 255 just after 58.
Dubious, test returned 255 (wstat 65280, 0xff00)
All 58 subtests passed 

Test Summary Report
-------------------
t/01-hash.t (Wstat: 65280 (exited 255) Tests: 58 Failed: 0)
  Non-zero exit status: 255
  Parse errors: No plan found in TAP output
Files=1, Tests=58,  0 wallclock secs ( 0.04 usr  0.01 sys +  0.09 cusr  0.02 csys =  0.16 CPU)
Result: FAIL
```

**Comment by richardleach** [comment]
> ... which means this has the same breakage point as the modules cited in #20114. @richardleach, can you take a look?

I'll take a look this evening.

> Given the amount of downstream breakage this change has generated, should we consider reverting it? (Or is it simply exposing dodgy code downstream?)

The downstream code for which I've supplied patches so far hasn't been dodgy as such, it's just been brittle in that it depends on optrees looking a certain way. Any optree optimizations that we introduce - I have a few more planned, LeoNerd has some planned - have a good chance of causing this kind of breakage. The only way to avoid breakage is just not to add optree optimizations, which - at least when we're talking about straightforward optimizations like this one - doesn't seem like the right balance to strike.

I hope that we're early enough in the release cycle to flush out all the affected modules and send patches upstream.

However, if there's a consensus that it should be reverted in a few months time, so be it.

**Comment by demerphq** [comment]
On Mon, 22 Aug 2022, 11:29 Richard Leach, ***@***.***> wrote:

> ... which means this has the same breakage point as the modules cited in
> #20114 <https://github.com/Perl/perl5/issues/20114>. @richardleach
> <https://github.com/richardleach>, can you take a look?
>
> I'll take a look this evening.
>
> Given the amount of downstream breakage this change has generated, should
> we consider reverting it? (Or is it simply exposing dodgy code downstream?)
>
> The downstream code for which I've supplied patches so far hasn't been
> dodgy as such, it's just been brittle in that it depends on optrees looking
> a certain way. Any optree optimizations that we introduce - I have a few
> more planned, LeoNerd has some planned - have a good chance of causing this
> kind of breakage. The only way to avoid breakage is just not to add optree
> optimizations, which - at least when we're talking about straightforward
> optimizations like this one - doesn't seem like the right balance to strike.
>
How very mildly put.  I'd have put that in much stronger terms. :-)


I hope that we're early enough in the release cycle to flush out all the
> affected modules and send patches upstream.
>
> However, if there's a consensus that it should be reverted in a few months
> time, so be it.
>

I really think that shouldn't happen. We have to be free to introduce new
opcodes and to deprecate and eliminate old ones.

Maybe we need an "opcode version" so that test files can at least
distinguish "things changed and my assumptions are out of date" from
"things changed and now stuff is broken".

Yves

>

**Comment by richardleach** [comment]
The Hash::MoreUtils breakage was down to a bug, which has been fixed in https://github.com/Perl/perl5/commit/6f629715866f9fa9452688fb66ee6661307b7a78

**Comment by jkeenan** [comment]
> The Hash::MoreUtils breakage was down to a bug, which has been fixed in [6f62971](https://github.com/Perl/perl5/commit/6f629715866f9fa9452688fb66ee6661307b7a78)

Confirmed by installing Hash-MoreUtils against a threaded perl built at v5.37.2-418-g8975221916.
